hal/network/ncp/wifi/platform_ncp_update.cpp

```diff
@@ -0,0 +1,167 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHAN'TABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform_ncp.h"
+#include "network/ncp/wifi/ncp.h"
+#include "network/ncp/wifi/wifi_network_manager.h"
+#include "network/ncp/wifi/wifi_ncp_client.h"
+#include "led_service.h"
+#include "check.h"
+#include "scope_guard.h"
+#include "system_led_signal.h"
+#include "stream.h"
+#include "debug.h"
+#include "ota_flash_hal_impl.h"
+
+#include <algorithm>
+
+class OtaUpdateSourceStream : public particle::InputStream {
+public:
+    OtaUpdateSourceStream(const uint8_t* buffer, size_t length) : buffer(buffer), remaining(length) {}
+
+    int read(char* data, size_t size) override {
+        CHECK(peek(data, size));
+        return skip(size);
+    }
+
+    int peek(char* data, size_t size) override {
+        if (!remaining) {
+            return SYSTEM_ERROR_END_OF_STREAM;
+        }
+        size = std::min(size, remaining);
+        memcpy(data, buffer, size);
+        return size;
+    }
+
+    int skip(size_t size) override {
+        if (!remaining) {
+            return SYSTEM_ERROR_END_OF_STREAM;
+        }
+        size = std::min(size, remaining);
+        buffer += size;
+        remaining -= size;
+        LED_Toggle(LED_RGB);
+        return size;
+    }
+
+    int availForRead() override {
+        return remaining;
+    }
+
+    int waitEvent(unsigned flags, unsigned timeout) override {
+        if (!flags) {
+            return 0;
+        }
+        if (!(flags & InputStream::READABLE)) {
+            return SYSTEM_ERROR_INVALID_ARGUMENT;
+        }
+        if (!remaining) {
+            return SYSTEM_ERROR_END_OF_STREAM;
+        }
+        return InputStream::READABLE;
+    }
+
+private:
+    const uint8_t* buffer;
+    size_t remaining;
+};
+
+// FIXME: This function accesses the module info via XIP and may fail to parse it correctly under
+// some not entirely clear circumstances. Disabling compiler optimizations helps to work around
+// the problem
+__attribute__((optimize("O0"))) int platform_ncp_update_module(const hal_module_t* module) {
+    const auto ncpClient = particle::wifiNetworkManager()->ncpClient();
+    SPARK_ASSERT(ncpClient);
+    CHECK(ncpClient->on());
+    // we pass only the actual binary after the module info and up to the suffix
+    const uint8_t* start = (const uint8_t*)module->info;
+    static_assert(sizeof(module_info_t)==24, "expected module info size to be 24");
```

Comment:
This seems like a hard magic number.  any way to derive this?

Reply:
If it's calculated, it removes the purpose of this check which is to ensure that a particular structure is indeed of a particular size.

Reply:
To clarify this is not a new code, it's https://github.com/particle-iot/device-os/blob/develop/hal/src/argon/platform_ncp_argon.cpp